flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/AggregateOperationFactory.java

```diff
@@ -501,16 +509,38 @@ protected Void defaultMethod(ResolvedExpression expression) {
 		}
 	}
 
-	private class GroupingExpressionValidator extends ResolvedExpressionDefaultVisitor<Void> {
-		@Override
-		protected Void defaultMethod(ResolvedExpression expression) {
-			TypeInformation<?> groupingType = expressionBridge.bridge(expression).resultType();
+	private static class IsKeyTypeChecker implements DataTypeVisitor<Boolean> {
 
-			if (!groupingType.isKeyType()) {
-				throw new ValidationException(format("Expression %s cannot be used as a grouping expression " +
-					"because it's not a valid key type which must be hashable and comparable", expression));
+		@Override
+		public Boolean visit(AtomicDataType atomicDataType) {
+			LogicalType logicalType = atomicDataType.getLogicalType();
+			if (logicalType instanceof LegacyTypeInformationType) {
+				return ((LegacyTypeInformationType) logicalType).getTypeInformation().isKeyType();
+			} else if (logicalType.getTypeRoot() == LogicalTypeRoot.ANY) {
+				// we don't know anything about the ANY type, we don't know if it is comparable and hashable.
+				return false;
+			} else if (logicalType instanceof StructuredType) {
+				StructuredType.StructuredComparision comparision = ((StructuredType) logicalType).getComparision();
+				return comparision == StructuredType.StructuredComparision.FULL ||
+					comparision == StructuredType.StructuredComparision.EQUALS;
 			}
-			return null;
+
+			return true;
+		}
+
+		@Override
+		public Boolean visit(CollectionDataType collectionDataType) {
+			return collectionDataType.getElementDataType().accept(this);
+		}
+
+		@Override
+		public Boolean visit(FieldsDataType fieldsDataType) {
+			return fieldsDataType.getFieldDataTypes().values().stream().allMatch(t -> t.accept(this));
+		}
+
+		@Override
+		public Boolean visit(KeyValueDataType keyValueDataType) {
+			return keyValueDataType.getKeyDataType().accept(this) && keyValueDataType.getValueDataType().accept(this);
```

Comment:
Do we really want to allow grouping on maps and arrays? Do systems like Oracle support this?

Reply:
I did some additional investigation for arrays:
* oracle does not really have ARRAY type. I checked for VARRAY and you cannot group on it.
* PostgreSQL has ARRAY type and you can group on it
* in SQL standard the only comment I could find says that if you try to group on user defined type with UNKNOWN comparison the behavior is implementation specific.

I think it should be fine to allow grouping on array types.

Reply:
Thanks for the investigation. Sounds good to me.